File: C++/ConvertSortedListToBinarySearchTree.cc
// Given a singly linked list where elements are sorted in ascending order,
// convert it to a height balanced BST.

class Solution {
    // precondition: first has more than 1 node
    ListNode* getMid(ListNode* first, ListNode* &second) {
        ListNode* fast = first->next->next;
        ListNode* slow = first;
        while (fast) {
            fast = fast->next;
            if (fast) {
                fast = fast->next;
                slow = slow->next;
            }
        }
        ListNode* mid = slow->next;
        slow->next = nullptr;
        second = mid->next;
        return mid;
    }
public:
    TreeNode* sortedListToBST(ListNode* head) {
        if (!head)
            return nullptr;
        if (!head->next)
            return new TreeNode(head->val);
        ListNode* second = nullptr;
        ListNode* mid = getMid(head, second);
        TreeNode* root = new TreeNode(mid->val);
        root->left = sortedListToBST(head);
        root->right = sortedListToBST(second);
        return root;
    }
};

File: Java/ConvertSortedListToBinarySearchTree.java
// Given a singly linked list where elements are sorted in ascending order, convert it to a height
// balanced BST.

public class ConvertSortedListToBinarySearchTree {
    public TreeNode sortedListToBST(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode h = new ListNode(0);
        h.next = head;
        ListNode slow = h;
        ListNode fast = head.next;
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
            if (fast != null) {
                fast = fast.next;
            }

        }
        ListNode next = slow.next;
        slow.next = null;
        TreeNode root = new TreeNode(next.val);
        root.left = sortedListToBST(h.next);
        root.right = sortedListToBST(next.next);
        return root;
    }
}
